### PR TITLE
Go: Add `go-autobuilder --identify-environment`

### PR DESCRIPTION
**Pull request by owen-mc**
Step 1 of https://github.com/github/codeql-go-team/issues/329.

Integration tests look tricky. I will instead use unit tests in `ql/go/extractor/cli/go-autobuilder/go-autobuilder_test.go` and some manual end-to-end testing. Once other languages have also implemented this flag we can set up everything up to make integration testing possible.

### CODE REVIEW

File: go/extractor/cli/go-autobuilder/go-autobuilder.go [hunk 1]
```diff
@@ -692,3 +695,161 @@ func main() {
...
+func outsideSupportedRange(version string) bool {
+	short := semver.MajorMinor("v" + version)
+	return semver.Compare(short, "v"+minGoVersion) < 0 || semver.Compare(short, "v"+maxGoVersion) > 0
+}
+
+func checkForUnsupportedVersions(v versionInfo) (msg, version string) {
+	if v.goDirectiveFound && outsideSupportedRange(v.goModVersion) {
+		msg = "The version of Go found in the `go.mod` file (" + v.goModVersion +
+			") is outside of the supported range (" + minGoVersion + "-" + maxGoVersion + ")."
+		version = ""
```
Comment (mbg) [inline go/extractor/cli/go-autobuilder/go-autobuilder.go:711]: Would it be better to return `nil` instead of an empty string if these functions cannot determine which go version to install? And have `msg` as an `error` rather than a `string`?

File: go/extractor/cli/go-autobuilder/go-autobuilder.go [hunk 2]
```diff
@@ -692,3 +695,161 @@ func main() {
...
+	}
+
+	msg, version = checkForVersionsNotFound(v)
+	if msg != "" {
+		return msg, version
+	}
+
+	msg, version = compareVersions(v)
+	return msg, version
+}
```
Comment (mbg) [inline go/extractor/cli/go-autobuilder/go-autobuilder.go:781]: The logic implemented in these functions is relatively complicated and tricky to follow / it is easy to miss corner cases. Once tests are added, those will increase confidence, but it would also be good to add documentation and more explicitly document assumptions/invariants throughout these functions.

File: go/extractor/cli/go-autobuilder/go-autobuilder.go [hunk 3]
```diff
@@ -692,3 +695,161 @@ func main() {
...
+		diagnostics.EmitUnsupportedVersionEnvironment(msg)
+	}
+
+	return msg, version
+}
+
+func checkForVersionsNotFound(v versionInfo) (msg, version string) {
+	if !v.goInstallationFound && !v.goDirectiveFound {
+		msg = "No version of Go installed and no `go.mod` file found. Writing an " +
+			"`environment.json` file specifying the maximum supported version of Go (" +
```
Comment (mbg) [inline go/extractor/cli/go-autobuilder/go-autobuilder.go:728]: For all the occurrences of `environment.json` here, you can probably use a generic name here.

**Review by mbg — CHANGES_REQUESTED** [review]

File: go/extractor/diagnostics/diagnostics.go [hunk 4]
```diff
@@ -192,3 +193,80 @@ func EmitRelativeImportPaths() {
...
+		noLocation,
+	)
+}
+
+func EmitNoGoModAndNoGoEnv(msg string) {
+	emitDiagnostic(
+		"go/identify-environment/no-go-mod-and-no-go-env",
+		"No `go.mod` file found and no Go version in environment",
+		msg,
+		severityNote,
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/extractor/diagnostics/diagnostics.go:224]: We will install the highest supported version of Go and attempt to build the project. There's a reasonable chance that it will work, I'd say.

File: go/extractor/cli/go-autobuilder/go-autobuilder.go (hunk 2 above)
Comment (owen-mc) [reply go/extractor/cli/go-autobuilder/go-autobuilder.go:781]: I'll add docstrings to the functions.

File: go/extractor/diagnostics/diagnostics.go [hunk 5]
```diff
@@ -192,3 +193,80 @@ func EmitRelativeImportPaths() {
 		noLocation,
 	)
 }
+
+func EmitUnsupportedVersionGoMod(msg string) {
+	emitDiagnostic(
+		"go/identify-environment/unsupported-version-in-go-mod",
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/extractor/diagnostics/diagnostics.go:199]: I've changed it so that `go/autobuilder/env-` is a common prefix to all the these new diagnostics.

File: go/extractor/diagnostics/diagnostics.go [hunk 6]
```diff
@@ -192,3 +193,80 @@ func EmitRelativeImportPaths() {
...
+		"Unsupported Go version in `go.mod` file",
+		msg,
+		severityError,
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/extractor/diagnostics/diagnostics.go:202]: I've changed them all to have severityNote. As they are telemetry-only it doesn't realy matter.

File: go/extractor/cli/go-autobuilder/go-autobuilder.go [hunk 7]
```diff
@@ -692,3 +695,161 @@ func main() {
...
+}
+
+func (v versionInfo) String() string {
+	return fmt.Sprintf("go.mod version: %s, go.mod directive found: %t, go env version: %s, go installation found: %t", v.goModVersion, v.goDirectiveFound, v.goEnvVersion, v.goInstallationFound)
+}
+
+func isGoInstalled() bool {
+	_, err := exec.LookPath("go")
+	return err == nil
+}
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/extractor/cli/go-autobuilder/go-autobuilder.go:825]: It's a well-known program so I'm happy to take that risk. We will fall over very soon if it isn't the go compiler because we'll try to parse the output of `go version`.

File: go/extractor/cli/go-autobuilder/go-autobuilder.go (hunk 1 above)
Comment (owen-mc) [reply go/extractor/cli/go-autobuilder/go-autobuilder.go:711]: I like the `nil` idea but strings can't be `nil` in go. I have effectively given the empty string a special meaning, and I've been careful to always assign a value to `version` even when I could leave it as the default value so that it's easier to spot any places where it hasn't been considered.

Comment (owen-mc) [reply go/extractor/cli/go-autobuilder/go-autobuilder.go:711]: I don't think `msg` should be an `error` because then people will assume that it's only non-`nil` if something has gone wrong, whereas it's actually non-`nil` on every path through the code.

**Review by mbg — APPROVED** [review]
Looks good to me for now
